src/main.cpp

```diff
@@ -2308,6 +2425,35 @@ bool ContextualCheckInputs(
                     return state.DoS(100,false, REJECT_INVALID, strprintf("mandatory-script-verify-flag-failed (%s)", ScriptErrorString(check.GetScriptError())));
                 }
             }
+
+            if (consensusParams.FeatureActive(nHeight, Consensus::ZIP222_TZE)) {
+                // for each TZE input, look up the associated prior TZE output, and pass both
+                // to the extension checker.
+                for (unsigned int i = 0; i < tx.vtzein.size(); i++) {
+                    const CTzeOutPoint& prevout = tx.vtzein[i].prevout;
+                    const CCoins* pCoins = inputs.AccessCoins(prevout.hash);
+                    assert(pCoins && pCoins->vtzeout.size() > prevout.n);
+
+                    // Check that the witness has the same tze type and mode as the
+                    // predicate. This might be duplicative of a check within the
+                    // TZE code itself?
+                    const CTzeData& witness = tx.vtzein[i].witness;
+                    const CTzeData& predicate = pCoins->vtzeout[prevout.n].first.predicate;
+                    if (witness.corresponds(predicate)) {
+                        // construct the context
+                        TzeContext ctx(nHeight, tx);
+
+                        // call through to the rust API's verify function
+                        if (!tze.check(consensusBranchId, predicate, witness, ctx)) {
+                            // FIXME: Need proper levels
+                            return state.DoS(10, false, REJECT_INVALID, "tze check failed");
```

Comment:
I think 10 is fine for now. Once we have actual TZEs we can figure out how to do the "check against old consensus branch ID" check.
```suggestion
                            return state.DoS(10, false, REJECT_INVALID, "tze-witness-verify-failed");
```